File: tests/algorithms/x/Java/machine_learning/gradient_boosting_classifier.java
public class Main {
    static class Stump {
        long feature;
        double threshold;
        double left;
        double right;
        Stump(long feature, double threshold, double left, double right) {
            this.feature = feature;
            this.threshold = threshold;
            this.left = left;
            this.right = right;
        }
        Stump() {}
        @Override public String toString() {
            return String.format("{'feature': %s, 'threshold': %s, 'left': %s, 'right': %s}", String.valueOf(feature), String.valueOf(threshold), String.valueOf(left), String.valueOf(right));
        }
    }

    static double[][] features;
    static double[] target;
    static Stump[] models_1;
    static double[] predictions;
    static double acc;

    static double exp_approx(double x) {
        double term = 1.0;
        double sum_1 = 1.0;
        long i_1 = 1;
        while (i_1 < 10) {
            term = term * x / (((Number)(i_1)).doubleValue());
            sum_1 = sum_1 + term;
            i_1 = i_1 + 1;
        }
        return sum_1;
    }

    static double signf(double x) {
        if (x >= 0.0) {
            return 1.0;
        }
        return -1.0;
    }

    static double[] gradient(double[] target, double[] preds) {
        long n = target.length;
        double[] residuals_1 = ((double[])(new double[]{}));
        long i_3 = 0;
        while (i_3 < n) {
            double t_1 = target[(int)(i_3)];
            double y_1 = preds[(int)(i_3)];
            double exp_val_1 = exp_approx(t_1 * y_1);
            double res_1 = -t_1 / (1.0 + exp_val_1);
            residuals_1 = ((double[])(java.util.stream.DoubleStream.concat(java.util.Arrays.stream(residuals_1), java.util.stream.DoubleStream.of(res_1)).toArray()));
            i_3 = i_3 + 1;
        }
        return residuals_1;
    }

    static double[] predict_raw(Stump[] models, double[][] features, double learning_rate) {
        long n_1 = features.length;
        double[] preds_1 = ((double[])(new double[]{}));
        long i_5 = 0;
        while (i_5 < n_1) {
            preds_1 = ((double[])(java.util.stream.DoubleStream.concat(java.util.Arrays.stream(preds_1), java.util.stream.DoubleStream.of(0.0)).toArray()));
            i_5 = i_5 + 1;
        }
        long m_1 = 0;
        while (m_1 < models.length) {
            Stump stump_1 = models[(int)(m_1)];
            i_5 = 0;
            while (i_5 < n_1) {
                double value_1 = features[(int)(i_5)][(int)(stump_1.feature)];
                if (value_1 <= stump_1.threshold) {
preds_1[(int)(i_5)] = preds_1[(int)(i_5)] + learning_rate * stump_1.left;
                } else {
preds_1[(int)(i_5)] = preds_1[(int)(i_5)] + learning_rate * stump_1.right;
                }
                i_5 = i_5 + 1;
            }
            m_1 = m_1 + 1;
        }
        return preds_1;
    }

    static double[] predict(Stump[] models, double[][] features, double learning_rate) {
        double[] raw = ((double[])(predict_raw(((Stump[])(models)), ((double[][])(features)), learning_rate)));
        double[] result_1 = ((double[])(new double[]{}));
        long i_7 = 0;
        while (i_7 < raw.length) {
            result_1 = ((double[])(java.util.stream.DoubleStream.concat(java.util.Arrays.stream(result_1), java.util.stream.DoubleStream.of(signf(raw[(int)(i_7)]))).toArray()));
            i_7 = i_7 + 1;
        }
        return result_1;
    }

    static Stump train_stump(double[][] features, double[] residuals) {
        long n_samples = features.length;
        long n_features_1 = features[(int)(0)].length;
        long best_feature_1 = 0;
        double best_threshold_1 = 0.0;
        double best_error_1 = 1000000000.0;
        double best_left_1 = 0.0;
        double best_right_1 = 0.0;
        long j_1 = 0;
        while (j_1 < n_features_1) {
            long t_index_1 = 0;
            while (t_index_1 < n_samples) {
                double t_3 = features[(int)(t_index_1)][(int)(j_1)];
                double sum_left_1 = 0.0;
                long count_left_1 = 0;
                double sum_right_1 = 0.0;
                long count_right_1 = 0;
                long i_9 = 0;
                while (i_9 < n_samples) {
                    if (features[(int)(i_9)][(int)(j_1)] <= t_3) {
                        sum_left_1 = sum_left_1 + residuals[(int)(i_9)];
                        count_left_1 = count_left_1 + 1;
                    } else {
                        sum_right_1 = sum_right_1 + residuals[(int)(i_9)];
                        count_right_1 = count_right_1 + 1;
                    }
                    i_9 = i_9 + 1;
                }
                double left_val_1 = 0.0;
                if (count_left_1 != 0) {
                    left_val_1 = sum_left_1 / (((Number)(count_left_1)).doubleValue());
                }
                double right_val_1 = 0.0;
                if (count_right_1 != 0) {
                    right_val_1 = sum_right_1 / (((Number)(count_right_1)).doubleValue());
                }
                double error_1 = 0.0;
                i_9 = 0;
                while (i_9 < n_samples) {
                    double pred_1 = features[(int)(i_9)][(int)(j_1)] <= t_3 ? left_val_1 : right_val_1;
                    double diff_1 = residuals[(int)(i_9)] - pred_1;
                    error_1 = error_1 + diff_1 * diff_1;
                    i_9 = i_9 + 1;
                }
                if (error_1 < best_error_1) {
                    best_error_1 = error_1;
                    best_feature_1 = j_1;
                    best_threshold_1 = t_3;
                    best_left_1 = left_val_1;
                    best_right_1 = right_val_1;
                }
                t_index_1 = t_index_1 + 1;
            }
            j_1 = j_1 + 1;
        }
        return new Stump(best_feature_1, best_threshold_1, best_left_1, best_right_1);
    }

    static Stump[] fit(long n_estimators, double learning_rate, double[][] features, double[] target) {
        Stump[] models = ((Stump[])(new Stump[]{}));
        long m_3 = 0;
        while (m_3 < n_estimators) {
            double[] preds_3 = ((double[])(predict_raw(((Stump[])(models)), ((double[][])(features)), learning_rate)));
            double[] grad_1 = ((double[])(gradient(((double[])(target)), ((double[])(preds_3)))));
            double[] residuals_3 = ((double[])(new double[]{}));
            long i_11 = 0;
            while (i_11 < grad_1.length) {
                residuals_3 = ((double[])(java.util.stream.DoubleStream.concat(java.util.Arrays.stream(residuals_3), java.util.stream.DoubleStream.of(-grad_1[(int)(i_11)])).toArray()));
                i_11 = i_11 + 1;
            }
            Stump stump_3 = train_stump(((double[][])(features)), ((double[])(residuals_3)));
            models = ((Stump[])(java.util.stream.Stream.concat(java.util.Arrays.stream(models), java.util.stream.Stream.of(stump_3)).toArray(Stump[]::new)));
            m_3 = m_3 + 1;
        }
        return models;
    }

    static double accuracy(double[] preds, double[] target) {
        long n_2 = target.length;
        long correct_1 = 0;
        long i_13 = 0;
        while (i_13 < n_2) {
            if (preds[(int)(i_13)] == target[(int)(i_13)]) {
                correct_1 = correct_1 + 1;
            }
            i_13 = i_13 + 1;
        }
        return (((Number)(correct_1)).doubleValue()) / (((Number)(n_2)).doubleValue());
    }
    public static void main(String[] args) {
        features = ((double[][])(new double[][]{new double[]{1.0}, new double[]{2.0}, new double[]{3.0}, new double[]{4.0}}));
        target = ((double[])(new double[]{-1.0, -1.0, 1.0, 1.0}));
        models_1 = ((Stump[])(fit(5, 0.5, ((double[][])(features)), ((double[])(target)))));
        predictions = ((double[])(predict(((Stump[])(models_1)), ((double[][])(features)), 0.5)));
        acc = accuracy(((double[])(predictions)), ((double[])(target)));
        System.out.println("Accuracy: " + _p(acc));
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        if (v instanceof Double || v instanceof Float) {
            double d = ((Number) v).doubleValue();
            if (d == Math.rint(d)) return String.valueOf((long) d);
            return String.valueOf(d);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/x/PHP/machine_learning/gradient_boosting_classifier.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
function _iadd($a, $b) {
    if (function_exists('bcadd')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return bcadd($sa, $sb, 0);
    }
    return $a + $b;
}
function _isub($a, $b) {
    if (function_exists('bcsub')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return bcsub($sa, $sb, 0);
    }
    return $a - $b;
}
function _imul($a, $b) {
    if (function_exists('bcmul')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return bcmul($sa, $sb, 0);
    }
    return $a * $b;
}
function _idiv($a, $b) {
    return _intdiv($a, $b);
}
function _imod($a, $b) {
    if (function_exists('bcmod')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcmod($sa, $sb));
    }
    return $a % $b;
}
$__start_mem = memory_get_usage();
$__start = _now();
  function exp_approx($x) {
  global $features, $target, $models, $predictions, $acc;
  $term = 1.0;
  $sum = 1.0;
  $i = 1;
  while ($i < 10) {
  $term = $term * $x / (floatval($i));
  $sum = $sum + $term;
  $i = _iadd($i, 1);
};
  return $sum;
};
  function signf($x) {
  global $features, $target, $models, $predictions, $acc;
  if ($x >= 0.0) {
  return 1.0;
}
  return -1.0;
};
  function gradient($target, $preds) {
  global $features, $models, $predictions, $acc;
  $n = count($target);
  $residuals = [];
  $i = 0;
  while ($i < $n) {
  $t = $target[$i];
  $y = $preds[$i];
  $exp_val = exp_approx($t * $y);
  $res = -$t / (1.0 + $exp_val);
  $residuals = _append($residuals, $res);
  $i = _iadd($i, 1);
};
  return $residuals;
};
  function predict_raw($models, $features, $learning_rate) {
  global $target, $predictions, $acc;
  $n = count($features);
  $preds = [];
  $i = 0;
  while ($i < $n) {
  $preds = _append($preds, 0.0);
  $i = _iadd($i, 1);
};
  $m = 0;
  while ($m < count($models)) {
  $stump = $models[$m];
  $i = 0;
  while ($i < $n) {
  $value = $features[$i][$stump['feature']];
  if ($value <= $stump['threshold']) {
  $preds[$i] = $preds[$i] + $learning_rate * $stump['left'];
} else {
  $preds[$i] = $preds[$i] + $learning_rate * $stump['right'];
}
  $i = _iadd($i, 1);
};
  $m = _iadd($m, 1);
};
  return $preds;
};
  function predict($models, $features, $learning_rate) {
  global $target, $predictions, $acc;
  $raw = predict_raw($models, $features, $learning_rate);
  $result = [];
  $i = 0;
  while ($i < count($raw)) {
  $result = _append($result, signf($raw[$i]));
  $i = _iadd($i, 1);
};
  return $result;
};
  function train_stump($features, $residuals) {
  global $target, $models, $predictions, $acc;
  $n_samples = count($features);
  $n_features = count($features[0]);
  $best_feature = 0;
  $best_threshold = 0.0;
  $best_error = 1000000000.0;
  $best_left = 0.0;
  $best_right = 0.0;
  $j = 0;
  while ($j < $n_features) {
  $t_index = 0;
  while ($t_index < $n_samples) {
  $t = $features[$t_index][$j];
  $sum_left = 0.0;
  $count_left = 0;
  $sum_right = 0.0;
  $count_right = 0;
  $i = 0;
  while ($i < $n_samples) {
  if ($features[$i][$j] <= $t) {
  $sum_left = $sum_left + $residuals[$i];
  $count_left = _iadd($count_left, 1);
} else {
  $sum_right = $sum_right + $residuals[$i];
  $count_right = _iadd($count_right, 1);
}
  $i = _iadd($i, 1);
};
  $left_val = 0.0;
  if ($count_left != 0) {
  $left_val = $sum_left / (floatval($count_left));
}
  $right_val = 0.0;
  if ($count_right != 0) {
  $right_val = $sum_right / (floatval($count_right));
}
  $error = 0.0;
  $i = 0;
  while ($i < $n_samples) {
  $pred = ($features[$i][$j] <= $t ? $left_val : $right_val);
  $diff = $residuals[$i] - $pred;
  $error = $error + $diff * $diff;
  $i = _iadd($i, 1);
};
  if ($error < $best_error) {
  $best_error = $error;
  $best_feature = $j;
  $best_threshold = $t;
  $best_left = $left_val;
  $best_right = $right_val;
}
  $t_index = _iadd($t_index, 1);
};
  $j = _iadd($j, 1);
};
  return ['feature' => $best_feature, 'threshold' => $best_threshold, 'left' => $best_left, 'right' => $best_right];
};
  function fit($n_estimators, $learning_rate, $features, $target) {
  global $predictions, $acc;
  $models = [];
  $m = 0;
  while ($m < $n_estimators) {
  $preds = predict_raw($models, $features, $learning_rate);
  $grad = gradient($target, $preds);
  $residuals = [];
  $i = 0;
  while ($i < count($grad)) {
  $residuals = _append($residuals, -$grad[$i]);
  $i = _iadd($i, 1);
};
  $stump = train_stump($features, $residuals);
  $models = _append($models, $stump);
  $m = _iadd($m, 1);
};
  return $models;
};
  function accuracy($preds, $target) {
  global $features, $models, $predictions, $acc;
  $n = count($target);
  $correct = 0;
  $i = 0;
  while ($i < $n) {
  if ($preds[$i] == $target[$i]) {
  $correct = _iadd($correct, 1);
}
  $i = _iadd($i, 1);
};
  return (floatval($correct)) / (floatval($n));
};
  $features = [[1.0], [2.0], [3.0], [4.0]];
  $target = [-1.0, -1.0, 1.0, 1.0];
  $models = fit(5, 0.5, $features, $target);
  $predictions = predict($models, $features, 0.5);
  $acc = accuracy($predictions, $target);
  echo rtrim('Accuracy: ' . _str($acc)), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
